### ISSUE TITLE
S3 API: Gap Analysis

### ISSUE THREAD

**Issue by diversemix** [libero/shared-services-experiment#32]
See if there is anything that is missing from the S3 API that would prevent us from:

  - [ ] Query by folder?
  - [ ] Query by tag?
  - [ ] Events related to file creation?

**Comment by giorgiosironi** [comment]
As far as I know:
- query by folder (`prefix` on S3) is the most common route to enumerate a small subset of a bucket
- query by tag (metadata in our case) is not possible in general, being a key-value store
- events can be implemented with [S3 notifications](https://docs.aws.amazon.com/AmazonS3/latest/dev/NotificationHowTo.html) which have a fixed schema

**Comment by diversemix** [comment]
Taking a look at the nodejs library: https://www.npmjs.com/package/minio

- You can do bucket notifications see: https://github.com/minio/minio-js/blob/master/examples/minio/listen-bucket-notification.js
- Query by folder - see comment by @giorgiosironi above.
- Query by tag however does seem to be a sticking point. We do want the Storage Service to be able to respond to questions like "Get all images from this article" or "Get all supporting files for this submission" ... and unless we can support this it pretty much a show stopper. The closest I have come to this is doing something like:
```
aws s3api head-object --bucket pdh-test2 --key README.md
```
Where `key` is a file - and doing this for all files in a folder does get you back the meta-data without needing to get back the entire file, something like:
```
{
    "AcceptRanges": "bytes",
    "LastModified": "Thu, 22 Aug 2019 16:58:43 GMT",
    "ContentLength": 28,
    "ETag": "\"e53beb3b953077d67bf4f12a4841fa41\"",
    "ContentType": "text/markdown",
    "Metadata": {
        "foo": "bar"
    }
}
```
Doing this for all files and then filtering in the calling code would seem both slow and repetitious. Happy to keep this question open - but before we go any further implementing S3 API's in client code - we need a solution to this. 

In short: I don't think its possible and it's a show stopper for using minio as the Storage Service.

/cc @hdrury1 @GiancarloFusiello

**Comment by GiancarloFusiello** [comment]
Another option is, you could use an intermediate table/service called Submissions, for example. This could store a reference to assets relating to a submission where those assets are stored in an object store.

**Comment by giorgiosironi** [comment]
> Doing this for all files and then filtering in the calling code would seem both slow and repetitious

You are right that this is a O(N) operation; the use cases look like that N is the number of files inside an Aggregate (like a Submission) rather than the set of all files. That puts the onus on the client to have another source of state to keep track of the tags to query it, like the proposed intermediate table. A Storage Service implements that table.
